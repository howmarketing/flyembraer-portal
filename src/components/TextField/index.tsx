import { forwardRef, ForwardRefRenderFunction, InputHTMLAttributes, useState } from "react"

import { CSS } from "@stitches/react"

import { Flex } from "components/Flex"
import { EyeOutline } from "components/Icons/EyeOutline"
import { EyeOffOutline } from "components/Icons/EyeOffOutline"

import { css, styled } from "../../../stitches.config"

type TextFieldProps = {
	id: string
	css?: CSS
	children?: string
	error?: string
	hint?: string
	required?: boolean
	isPassword?: boolean
} & InputHTMLAttributes<HTMLInputElement>

const TextField: ForwardRefRenderFunction<HTMLInputElement, TextFieldProps> = (
	{ id, css, children, hint, error, type = "text", required = false, isPassword = false, ...props },
	ref
) => {
	const [showPassword, setShowPassword] = useState(false)

	function handleShowPassword() {
		setShowPassword((prevState) => !prevState)
	}

	return (
		<div className={styledWrapper({ css })}>
			{children && <StyledLabel htmlFor={id}>{required ? `${children} *` : `${children}`}</StyledLabel>}
			{error && <StyledErrorMessage id={`${id}_error_message`}>{error}</StyledErrorMessage>}

			<Flex css={{ position: "relative" }}>
				<StyledInput
					id={id}
					name={id}
					type={showPassword ? "text" : type}
					aria-required={required}
					aria-invalid={error ? true : false}
					aria-describedby={error ? `${id}_error_message` : undefined}
					ref={ref}
					{...props}
				/>

				{isPassword &&
					(showPassword ? (
						<PasswordButton type="button" onClick={handleShowPassword}>
							<EyeOutline width="22" height="18" />
						</PasswordButton>
					) : (
						<PasswordButton type="button" onClick={handleShowPassword}>
							<EyeOffOutline width="22" height="18" />
						</PasswordButton>
					))}
			</Flex>

			{!error && hint ? <StyledHint>{hint}</StyledHint> : null}
		</div>
	)
}

export default forwardRef(TextField)

const styledWrapper = css({
	width: "100%",
	display: "flex",
	flexDirection: "column",
	gap: "$4",
})

const StyledLabel = styled("label", {
	fontSize: "$sm",
	lineHeight: "$1",
	fontWeight: "$thin",
	color: "$black200",
})

const StyledErrorMessage = styled("p", {
	fontSize: "$sm",
	lineHeight: "14px",
	color: "$error",
})

const StyledInput = styled("input", {
	width: "100%",
	height: "40px",

	display: "inline-block",
	px: "$16",

	color: "$gray300",
	background: "$white",
	borderRadius: "4px",
	border: "1px solid $gray100",
	boxShadow: "inset 0px 1px 2px rgba(0, 0, 0, 0.15)",

	fontSize: "$sm",

	"&:disabled": {
		background: "$gray100",
		cursor: "not-allowed",
	},
})

const StyledHint = styled("p", {
	fontWeight: "$thin",
	fontSize: "$sm",
	lineHeight: "14px",
	color: "$gray300",
})

const PasswordButton = styled("button", {
	all: "unset",
	appearance: "none",
	width: "min-content",
	height: "auto",
	p: "0.25rem",
	borderRadius: "100%",
	bg: "$white",
	position: "absolute",
	top: "50%",
	right: "12px",
	transform: "translateY(-50%)",
	cursor: "pointer",

	"&:focus": {
		boxShadow: "0 0 0 1px rgba(0, 0, 0, 0.7)",
	},
})
