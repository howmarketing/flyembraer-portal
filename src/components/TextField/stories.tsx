import { ComponentMeta, ComponentStory } from "@storybook/react"

import TextField from "."

export default {
	title: "Components/TextField",
	component: TextField,
	args: {
		id: "label",
		children: "Text field",
	},
} as ComponentMeta<typeof TextField>

const Template: ComponentStory<typeof TextField> = (args) => <TextField {...args}>{args.children}</TextField>

export const Default = Template.bind({})

export const Required = Template.bind({})
Required.args = {
	required: true,
}

export const WithError = Template.bind({})
WithError.args = {
	error: "This field is required",
	required: true,
}

export const Disabled = Template.bind({})
Disabled.args = {
	disabled: true,
}
