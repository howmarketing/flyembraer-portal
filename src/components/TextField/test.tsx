import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import TextField from "."

describe("<TextField />", () => {
	it("should render the TextField and match the previous version", () => {
		const { container } = render(<TextField id="label">Label</TextField>)

		expect(screen.queryByLabelText("Label")).toBeInTheDocument()

		expect(container.firstChild).toMatchSnapshot()
	})

	it("should show the field is required", () => {
		render(
			<TextField id="label" required>
				Label
			</TextField>
		)

		expect(screen.queryByLabelText("Label *")).toBeInTheDocument()
	})

	it("should render with an error", () => {
		render(
			<TextField id="label" error="Error">
				Label
			</TextField>
		)

		expect(screen.queryByText("Error")).toBeInTheDocument()
	})

	it("should be accessible by tab", () => {
		render(<TextField id="label">Label</TextField>)

		const input = screen.getByLabelText("Label")
		expect(document.body).toHaveFocus()

		userEvent.tab()
		expect(input).toHaveFocus()
	})
})
