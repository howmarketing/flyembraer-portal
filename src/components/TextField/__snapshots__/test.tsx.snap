// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render the TextField and match the previous version 1`] = `
<div
  class="c-fGJxyJ"
>
  <label
    class="c-jwfAEw"
    for="label"
  >
    Label
  </label>
  <input
    aria-invalid="false"
    aria-required="false"
    class="c-fsJYtE"
    id="label"
    name="label"
    type="text"
  />
</div>
`;
